src/coreclr/jit/morph.cpp

```diff
@@ -293,6 +293,41 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
     var_types dstType = tree->CastToType();
     unsigned  dstSize = genTypeSize(dstType);
 
+#if defined(TARGET_AMD64)
+    // If AVX512 is present, we have intrinsic available to convert
+    // ulong directly to float. Hence, we need to combine the 2 nodes
+    // GT_CAST(GT_CAST(TYP_ULONG, TYP_DOUBLE), TYP_FLOAT) into a single
+    // node i.e. GT_CAST(TYP_ULONG, TYP_FLOAT). At this point, we already
+    // have the 2 GT_CAST nodes in the tree and we are combining them below.
+    if (compOpportunisticallyDependsOn(InstructionSet_AVX512F))
```

Comment:
Same here, we should do the relevant checks and then if we're in a scenario where AVX512 would actually be used, do the opportunistic query.

This makes a difference for R2R scenarios where the dev has opted into AVX-512 as the ISA will return `true` for opportunistic support and then the method will get marked as using AVX-512.